cpp/src/scalar/scalar.cpp

```diff
@@ -56,4 +95,352 @@ std::string string_scalar::to_string(rmm::cuda_stream_view stream) const
+}
+
+template <typename T>
+typename fixed_point_scalar<T>::rep_type* fixed_point_scalar<T>::data()
+{
+  return _data.data();
+}
+
+template <typename T>
+typename fixed_point_scalar<T>::rep_type const* fixed_point_scalar<T>::data() const
+{
+  return _data.data();
+}
+
+template class fixed_point_scalar<numeric::decimal32>;
+template class fixed_point_scalar<numeric::decimal64>;
+
+namespace detail {
+
+template <typename T>
+fixed_width_scalar<T>::fixed_width_scalar() : scalar(data_type(type_to_id<T>()))
+{
+}
+
+template <typename T>
+fixed_width_scalar<T>::fixed_width_scalar(T value,
+                                          bool is_valid,
+                                          rmm::cuda_stream_view stream,
+                                          rmm::mr::device_memory_resource* mr)
+  : scalar(data_type(type_to_id<T>()), is_valid, stream, mr), _data(value, stream, mr)
+{
+}
+
+template <typename T>
+fixed_width_scalar<T>::fixed_width_scalar(rmm::device_scalar<T>&& data,
+                                          bool is_valid,
+                                          rmm::cuda_stream_view stream,
+                                          rmm::mr::device_memory_resource* mr)
+  : scalar(data_type(type_to_id<T>()), is_valid, stream, mr),
+    _data{std::forward<rmm::device_scalar<T>>(data)}
+{
+}
+
+template <typename T>
+void fixed_width_scalar<T>::set_value(T value, rmm::cuda_stream_view stream)
+{
+  _data.set_value(value, stream);
+  this->set_valid(true, stream);
+}
+
+template <typename T>
+T fixed_width_scalar<T>::value(rmm::cuda_stream_view stream) const
+{
+  return _data.value(stream);
+}
+
+template <typename T>
+T* fixed_width_scalar<T>::data()
+{
+  return _data.data();
+}
+
+template <typename T>
+T const* fixed_width_scalar<T>::data() const
+{
+  return _data.data();
+}
+
+template <typename T>
+fixed_width_scalar<T>::operator value_type() const
+{
+  return this->value(rmm::cuda_stream_default);
+}
+
+template class fixed_width_scalar<bool>;
+template class fixed_width_scalar<int8_t>;
+template class fixed_width_scalar<int16_t>;
+template class fixed_width_scalar<int32_t>;
+template class fixed_width_scalar<int64_t>;
+template class fixed_width_scalar<uint8_t>;
+template class fixed_width_scalar<uint16_t>;
+template class fixed_width_scalar<uint32_t>;
+template class fixed_width_scalar<uint64_t>;
+template class fixed_width_scalar<float>;
+template class fixed_width_scalar<double>;
+template class fixed_width_scalar<timestamp_D>;
+template class fixed_width_scalar<timestamp_s>;
+template class fixed_width_scalar<timestamp_ms>;
+template class fixed_width_scalar<timestamp_us>;
+template class fixed_width_scalar<timestamp_ns>;
+template class fixed_width_scalar<duration_D>;
+template class fixed_width_scalar<duration_s>;
+template class fixed_width_scalar<duration_ms>;
+template class fixed_width_scalar<duration_us>;
+template class fixed_width_scalar<duration_ns>;
```

Comment:
I would like to use a macro to iterate through all types in a loop at compile time? Something similar to this:
```
#define INSTANTIATE(Type, ...)  \
    template class fixed_width_scalar(Type); \
    INSTANTIATE(__VA_ARGS__)

INSTANTIATE(int8_t, int16_t, ........................, duration_ns)
```

Reply:
I disagree with this recommendation. 
I think the code here is simpler to understand and maintain as is.